ci/cron/src/Main.hs

```diff
@@ -113,34 +113,77 @@ to_v s = case map read $ Split.splitOn "." s of
-    let cur_sha = init out -- remove ending newline
-    shell_ $ "mkdir -p " <> path
-    let latest = head versions
-    putStrLn $ "Building latest docs: " <> latest
-    shell_ $ "git checkout v" <> latest
-    robustly_download_nix_packages
-    shell_ "bazel build //docs:docs"
-    shell_ $ "tar xzf bazel-bin/docs/html.tar.gz --strip-components=1 -C " <> path
-    -- Not going through Aeson because it represents JSON objects as unordered
-    -- maps, and here order matters.
-    let versions_json = versions
-                        & map (\s -> "\"" <> s <> "\": \"" <> s <> "\"")
-                        & List.join ", "
-                        & \s -> "{" <> s <> "}"
-    writeFile (path <> "/versions.json") versions_json
-    shell_ $ "mkdir -p  " <> path <> "/" <> latest
-    shell_ $ "tar xzf bazel-bin/docs/html.tar.gz --strip-components=1 -C " <> path <> "/" <> latest
-    Foldable.for_ (tail versions) $ \version -> do
-        putStrLn $ "Building older docs: " <> version
-        shell_ $ "git checkout v" <> version
-        robustly_download_nix_packages
-        shell_ "bazel build //docs:docs"
-        shell_ $ "mkdir -p  " <> path <> "/" <> version
-        shell_ $ "tar xzf bazel-bin/docs/html.tar.gz --strip-components=1 -C" <> path <> "/" <> version
-    shell_ $ "git checkout " <> cur_sha
+build_docs_folder :: String -> [String] -> String -> IO String
+build_docs_folder path versions latest = do
+    restore_sha $ do
+        let old = path <> "/old"
+        let new = path <> "/new"
+        download_existing_site_from_s3 old
+        Foldable.for_ versions $ \version -> do
+            putStrLn $ "Building " <> version <> "..."
+            putStrLn "  Checking for existing folder..."
+            old_version_exists <- exists old version
+            if old_version_exists
+            then do
+                -- Note: this checks for upload errors; this is NOT in any way
+                -- a protection against tampering at the s3 level as we get the
+                -- checksums from the s3 bucket.
+                putStrLn "  Found. Checking integrity..."
+                checksums_match <- checksums old version
+                if checksums_match
+                then do
+                    putStrLn "  Checks, reusing existing."
+                    copy (old <> "/" <> version) $ new <> "/" <> version
+                else do
+                    putStrLn "  Check failed. Rebuilding..."
+                    build version new
+            else do
+                putStrLn "  Not found. Building..."
+                build version new
+            putStrLn $ "Done " <> version <> "."
+        putStrLn $ "Copying latest (" <> latest <> ") to top-level..."
+        copy (new <> "/" <> latest <> "/*") (new <> "/")
+        putStrLn "Creating versions.json..."
+        create_versions_json versions new
+        return new
+    where
+        restore_sha io = do
+            cur_sha <- init <$> shell "git rev-parse HEAD"
```

Comment:
You probably want to use `bracket` here to restore the old sha in the case of an exception. Not super important here probably since you will just crash the whole program but it’s generally good practice and imho also makes the intent clearer.

Reply:
Thanks, I did not know about bracket. I assume that's `Control.Exception.bracket`?

Reply:
exactly!